buffer/src/main/java/io/netty/buffer/AbstractReadableObject.java

```diff
@@ -0,0 +1,187 @@
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.buffer;
+
+import io.netty.util.IllegalReferenceCountException;
+import io.netty.util.ResourceLeak;
+import io.netty.util.ResourceLeakDetector;
+
+import java.io.IOException;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * A skeletal implementation of a {@link ReadableObject}.
+ */
+public abstract class AbstractReadableObject implements ReadableObject {
+    private static final ResourceLeakDetector<ReadableObject> leakDetector =
+            new ResourceLeakDetector<ReadableObject>(ReadableObject.class);
+
+    private final ResourceLeak leak;
+    private long readerPosition;
+
+    public AbstractReadableObject(long readerPosition) {
+        this.readerPosition = readerPosition;
+        leak = leakDetector.open(this);
+    }
+
+    @Override
+    public final long readerPosition() {
+        return readerPosition;
+    }
+
+    @Override
+    public final long readableBytes() {
+        return readerLimit() - readerPosition;
+    }
+
+    @Override
+    public ReadableObject readerPosition(long readerPosition) {
+        if (readerPosition < 0 || readerPosition > readerLimit()) {
+            throw new IndexOutOfBoundsException("readerPosition " + readerPosition);
+        }
+        this.readerPosition = readerPosition;
+        return this;
+    }
+
+    @Override
+    public final boolean isReadable() {
+        return readableBytes() > 0;
+    }
+
+    @Override
+    public final boolean isReadable(int size) {
+        return readableBytes() >= size;
+    }
+
+    @Override
+    public ReadableObject skipBytes(long length) {
+        ensureReadable(length);
+        readerPosition += length;
+        return this;
+    }
+
+    @Override
+    public final long readTo(WritableByteChannel target, long pos, long length) throws IOException {
+        if (refCnt() == 0) {
+            throw new IllegalReferenceCountException(0);
+        }
+        if (pos < 0 || length < 0 || readerLimit() - length < pos) {
+            throw new IndexOutOfBoundsException(String.format("pos: %d, length: %d", pos, length));
+        }
+        if (length == 0) {
+            return 0L;
+        }
+
+        return readTo0(target, pos, length);
+    }
+
+    @Override
+    public final ReadableObject slice() {
+        return slice(readerPosition, readableBytes());
+    }
+
+    @Override
+    public ReadableObject slice(long position, long length) {
+        ensureReadable(length);
```

Comment:
Is this necessary? Shouldn't we just cover this in the next if statement?  We should be considering the `position` parameter when checking validity too.